[Power Launcher] Integrate web search experience powered by Bing apis

peterpy8 opened the pull request:
## Summary of the Pull Request

~~https://github.com/microsoft/PowerToys/issues/5119~~ #3245

## PR Checklist
* [ ] Applies to #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

A new plugin Microsoft.Plugin.WebSearch

## Validation Steps Performed

type 'web xxx' or 'map 1 microsoft way' or 'dict apple' or 'image titanic'.

niels9001 commented:
Thanks for the PR @peterpy8 .

For all other plugin images we use 48x48 dimensions, with a padding of 6px for top/bottom.

I have updated your assets to reflect this (and updated the Bing logo since it was using an older version of that) you can easily include:

![baidu](https://user-images.githubusercontent.com/9866362/89166569-89a15700-d57a-11ea-88d2-1fbcfd697a07.png)
![google](https://user-images.githubusercontent.com/9866362/89166571-8a39ed80-d57a-11ea-88b1-ca92d2affae6.png)
![bing](https://user-images.githubusercontent.com/9866362/89166572-8ad28400-d57a-11ea-8c75-2fe10478ba2f.png)

Since Bing Maps doesn't have an official logo, I'd just re-use the Bing one. Do you agree?

htcfreek commented:
Achtung: Bing bekommt wohl demnächst ein neues Logo.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.WebSearch/Languages/de.xaml:
```diff
@@ -0,0 +1,33 @@
+                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+                    xmlns:system="clr-namespace:System;assembly=mscorlib">
+
+    <system:String x:Key="wox_plugin_websearch_delete">Löschen</system:String>
```
Should this be changed to microsoft_plugin... ?

peterpy8 replied:
Good catch. I will change it.

peterpy8 commented:
> Thanks for the PR @peterpy8 .
> 
> For all other plugin images we use 48x48 dimensions, with a padding of 6px for top/bottom.
> 
> I have updated your assets to reflect this (and updated the Bing logo since it was using an older version of that) you can easily include:
> 
> ![baidu](https://user-images.githubusercontent.com/9866362/89166569-89a15700-d57a-11ea-88d2-1fbcfd697a07.png)
> ![google](https://user-images.githubusercontent.com/9866362/89166571-8a39ed80-d57a-11ea-88b1-ca92d2affae6.png)
> ![bing](https://user-images.githubusercontent.com/9866362/89166572-8ad28400-d57a-11ea-8c75-2fe10478ba2f.png)
> 
> Since Bing Maps doesn't have an official logo, I'd just re-use the Bing one. Do you agree?

Thanks for pointing this out. I will update the logo in my next iteration.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.WebSearch/Main.cs:
```diff
@@ -0,0 +1,198 @@
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Controls;
+using Wox.Infrastructure;
```
Why using wox classes?

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.WebSearch/Microsoft.Plugin.WebSearch.csproj:
```diff
@@ -0,0 +1,132 @@
+    </Page>
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\Wox.Infrastructure\Wox.Infrastructure.csproj" />
```
References to wox project. Not shure if this is correct.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.WebSearch/SearchSource.cs:
```diff
@@ -0,0 +1,42 @@
+using System.Windows.Media;
+using JetBrains.Annotations;
+using Newtonsoft.Json;
+using Wox.Plugin.WebSearch.SuggestionSources;
```
Wox namespace. Not shure if this is correct.

crutkas commented:
@peterpy8 is this just a modified version of https://github.com/peterpy8/Wox/tree/master/Plugins/Wox.Plugin.WebSearch?

peterpy8 commented:
> @peterpy8 is this just a modified version of https://github.com/peterpy8/Wox/tree/master/Plugins/Wox.Plugin.WebSearch?
> 
> https://github.com/Wox-launcher/Wox/tree/master/Plugins/Wox.Plugin.WebSearch

yes.

htcfreek commented:
@peterpy8 
Can please you provide/add an option to disable web search for people who don't want to search the internet.

crutkas commented:
I think this plugin, we need to start thinking about enabling / disabling plugins holistically.

Also this plug in needs to not break everything if the user does not have internet access.

htcfreek commented:
> I think this plugin, we need to start thinking about enabling / disabling plugins holistically.

@crutkas 
See #5273 for a concept/mock up about plugin management. 😃

peterpy8 commented:
> > I think this plugin, we need to start thinking about enabling / disabling plugins holistically.
> 
> @crutkas
> See #5273 for a concept/mock up about plugin management. 😃

I agree with crutkas's point that if no internet connectivity, the plugin should be disabled. Could you suggest a proper way to detect internet connectivity?

muzam1l commented:
Don't get me wrong, but bing is the primary reason for many to not use windows search who just want to search for files/programs on their computer. So instead of using so many generic words like 'web', 'map' or 'image' for bing search, which can be typed unintentionally by many ( like who just wanted to search for 'maps' app), you could have this plugin prefixed (in addition of option to disable) by some a symbol (like ~ might be good option).

crutkas commented:
Our goal with PowerToys Run was to build a quick way people can launch apps and files.  In doing so, our interface is a plug-in based one to enable all the core scenarios we needed to meet this need. When the plugin system is ready for 3rd party plugins, this will anyone to build against the APIs for their feature.  This work includes a robust settings UI that would allow someone to not load a plugin as well as set an action key.

The first main step of that is https://github.com/microsoft/PowerToys/issues/5273.
 
As of now, yes we understand some would want to be able to do a search from their instance of PT Run but others wouldn’t. As this isn’t a core functionality goal of PT Run currently, we’re going to decline this PR as we do need additional work to enable this for everyone to have a great experience.
